File: app/src/main/java/com/example/halanchallenge/LoginActivity.java
package com.example.halanchallenge;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

public class LoginActivity extends AppCompatActivity {

    EditText userNameEt, passwordEt;
    Button loginBtn;
    Login login;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        userNameEt = findViewById(R.id.et_username);
        passwordEt = findViewById(R.id.et_password);
        loginBtn = findViewById(R.id.btn_login);

        login = new Login();


        loginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                login.login(userNameEt.getText().toString(), passwordEt.getText().toString(),getApplicationContext());
            }
        });

    }
}

File: app/src/main/java/com/example/halanchallenge/ui/login/activities/LoginActivity.kt
package com.example.halanchallenge.ui.login.activities

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import com.example.halanchallenge.data.local.SharedPreferencesManager
import com.example.halanchallenge.databinding.ActivityLoginBinding
import com.example.halanchallenge.model.LoginCredentials
import com.example.halanchallenge.ui.login.viewmodels.LoginViewModel
import com.example.halanchallenge.ui.login.viewmodels.LoginViewState
import com.example.halanchallenge.ui.productslist.activities.ProductListActivity
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel


class LoginActivity : AppCompatActivity() {
    private lateinit var binding: ActivityLoginBinding
    private val viewModel: LoginViewModel by viewModel()
    private val sharedPref: SharedPreferencesManager by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initializeClickListeners()
        initializeObservers()

    }

    private fun initializeObservers() {
        viewModel.userData.observe(this, {
            renderUi(it)
        })
    }

    private fun renderUi(viewState: LoginViewState) {
        binding.apply {
            when (viewState) {
                is LoginViewState.LOADING -> pbLoading.isVisible = true
                is LoginViewState.SUCCESS -> {
                    pbLoading.isVisible = false
                    sharedPref.saveToken(viewState.payload.token)

                    startActivity(
                        Intent(
                            this@LoginActivity,
                            ProductListActivity::class.java
                        ).apply {
                            putExtra("Profile", viewState.payload.profile)
                        })
                    finish()
                }

                is LoginViewState.FAILURE -> {
                    pbLoading.isVisible = false
                    Toast.makeText(
                        this@LoginActivity,
                        viewState.errorMsg,
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        }


    }

    private fun initializeClickListeners() {
        binding.apply {
            btnLogin.setOnClickListener {
                viewModel.login(
                    LoginCredentials(
                        userName = etUsername.text.toString(),
                        password = etPassword.text.toString()
                    )
                )
            }
        }
    }
}